File: claude-code-agents-typescript/src/orchestrator/headless-runner.ts
// src/orchestrator/headless-runner.ts — HeadlessRunner wrapping claude -p subprocess
import { execSync, spawn } from 'node:child_process';
import type { Result } from '../types.js';
import { Err, Ok } from '../types.js';

// ── Runner Errors ───────────────────────────────────────────────
export class RunnerError extends Error {
  constructor(
    message: string,
    readonly exitCode?: number,
  ) {
    super(message);
    this.name = 'RunnerError';
  }
}

export class RunnerTimeoutError extends RunnerError {
  constructor(timeoutMs: number) {
    super(`HeadlessRunner timed out after ${timeoutMs}ms`);
    this.name = 'RunnerTimeoutError';
  }
}

export class RunnerNotFoundError extends RunnerError {
  constructor(binary: string) {
    super(
      `Claude binary not found: ${binary}. Ensure Claude CLI is installed and in PATH.`,
    );
    this.name = 'RunnerNotFoundError';
  }
}

// ── Configuration ───────────────────────────────────────────────
export interface HeadlessRunnerConfig {
  readonly claudeBinary: string;
  readonly timeoutMs: number;
  readonly model: string | undefined;
  readonly allowedTools: readonly string[] | undefined;
  readonly maxTurns: number;
}

const DEFAULT_CONFIG: HeadlessRunnerConfig = {
  claudeBinary: 'claude',
  timeoutMs: 300_000,
  model: undefined,
  allowedTools: undefined,
  maxTurns: 10,
};

// ── Stream Event ────────────────────────────────────────────────
export interface StreamEvent {
  readonly type: string;
  readonly [key: string]: unknown;
}

// ── HeadlessRunner ──────────────────────────────────────────────
export class HeadlessRunner {
  private readonly config: HeadlessRunnerConfig;

  constructor(config?: Partial<HeadlessRunnerConfig>) {
    this.config = { ...DEFAULT_CONFIG, ...config };
  }

  private buildCommand(prompt: string, streaming: boolean): readonly string[] {
    const cmd: string[] = [this.config.claudeBinary, '-p', prompt];
    if (streaming) {
      cmd.push('--output-format', 'stream-json');
    } else {
      cmd.push('--output-format', 'json');
    }
    if (this.config.model) {
      cmd.push('--model', this.config.model);
    }
    if (this.config.allowedTools) {
      for (const tool of this.config.allowedTools) {
        cmd.push('--allowedTools', tool);
      }
    }
    cmd.push('--max-turns', String(this.config.maxTurns));
    return cmd;
  }

  run(prompt: string): Result<string, RunnerError> {
    const cmd = this.buildCommand(prompt, false);
    const execResult = this.execSafe(cmd);
    if (!execResult.ok) return execResult;
    return Ok(this.parseJsonOutput(execResult.value));
  }

  private execSafe(cmd: readonly string[]): Result<string, RunnerError> {
    try {
      const result = execSync(cmd.join(' '), {
        timeout: this.config.timeoutMs,
        encoding: 'utf-8',
        stdio: ['pipe', 'pipe', 'pipe'],
      });
      return Ok(result);
    } catch (e: unknown) {
      if (e instanceof Error && 'killed' in e) {
        return Err(new RunnerTimeoutError(this.config.timeoutMs));
      }
      if (
        e instanceof Error &&
        'code' in e &&
        (e as NodeJS.ErrnoException).code === 'ENOENT'
      ) {
        return Err(new RunnerNotFoundError(this.config.claudeBinary));
      }
      const msg = e instanceof Error ? e.message : String(e);
      return Err(new RunnerError(`HeadlessRunner failed: ${msg}`));
    }
  }

  private parseJsonOutput(raw: string): string {
    const parsed = this.tryParseJson(raw);
    if (parsed !== undefined && typeof parsed['result'] === 'string') {
      return parsed['result'];
    }
    return raw.trim();
  }

  private tryParseJson(raw: string): Record<string, unknown> | undefined {
    try {
      return JSON.parse(raw) as Record<string, unknown>;
    } catch {
      return undefined;
    }
  }

  async runStreaming(
    prompt: string,
  ): Promise<Result<readonly StreamEvent[], RunnerError>> {
    const cmd = this.buildCommand(prompt, true);
    return new Promise((resolve) => {
      try {
        const proc = spawn(cmd[0]!, cmd.slice(1), {
          stdio: ['pipe', 'pipe', 'pipe'],
        });

        const events: StreamEvent[] = [];
        let stderr = '';

        proc.stdout.on('data', (chunk: Buffer) => {
          const lines = chunk.toString('utf-8').split('\n');
          for (const line of lines) {
            const trimmed = line.trim();
            if (!trimmed) continue;
            const parsed = this.tryParseJson(trimmed);
            if (parsed !== undefined) {
              events.push(parsed as StreamEvent);
            }
          }
        });

        proc.stderr.on('data', (chunk: Buffer) => {
          stderr += chunk.toString('utf-8');
        });

        proc.on('close', (code) => {
          if (code !== 0) {
            resolve(
              Err(
                new RunnerError(
                  `HeadlessRunner (streaming) exited with code ${code}: ${stderr.slice(0, 500)}`,
                  code ?? undefined,
                ),
              ),
            );
          } else {
            resolve(Ok(events));
          }
        });

        proc.on('error', (err) => {
          if ((err as NodeJS.ErrnoException).code === 'ENOENT') {
            resolve(Err(new RunnerNotFoundError(this.config.claudeBinary)));
          } else {
            resolve(Err(new RunnerError(err.message)));
          }
        });

        setTimeout(() => {
          proc.kill();
          resolve(Err(new RunnerTimeoutError(this.config.timeoutMs)));
        }, this.config.timeoutMs);
      } catch (e) {
        resolve(
          Err(
            new RunnerError(e instanceof Error ? e.message : String(e)),
          ),
        );
      }
    });
  }

  checkAvailable(): Result<boolean, RunnerError> {
    const result = this.execSafe([this.config.claudeBinary, '--version']);
    return result.ok ? Ok(true) : Err(new RunnerNotFoundError(this.config.claudeBinary));
  }
}

File: claude-code-agents-python/src/orchestrator/headless_runner.py
"""HeadlessRunner wrapping claude -p subprocess for Agent SDK execution."""

from __future__ import annotations

import json
import logging
import subprocess
from typing import Any, Optional

logger = logging.getLogger(__name__)


class HeadlessRunner:
    """Wraps `claude -p` subprocess calls for headless agent execution.

    This runner sends prompts to the Claude CLI in headless mode and
    captures the structured output. It supports both single-shot and
    streaming JSON output modes.

    Usage:
        runner = HeadlessRunner()
        output = runner.run("Crawl https://example.com and extract all links")
        events = runner.run_streaming("Analyze this codebase")
    """

    def __init__(
        self,
        claude_binary: str = "claude",
        timeout: int = 300,
        model: Optional[str] = None,
        allowed_tools: Optional[list[str]] = None,
        max_turns: int = 10,
    ) -> None:
        """Initialize the HeadlessRunner.

        Args:
            claude_binary: Path to the claude CLI binary.
            timeout: Timeout in seconds for each run.
            model: Model override (e.g., 'claude-sonnet-4-20250514').
            allowed_tools: List of tools to allow (e.g., ['Bash', 'Read']).
            max_turns: Maximum conversation turns.
        """
        self.claude_binary = claude_binary
        self.timeout = timeout
        self.model = model
        self.allowed_tools = allowed_tools
        self.max_turns = max_turns

    def _build_command(self, prompt: str, streaming: bool = False) -> list[str]:
        """Build the claude CLI command."""
        cmd = [self.claude_binary, "-p", prompt]
        if streaming:
            cmd.extend(["--output-format", "stream-json"])
        else:
            cmd.extend(["--output-format", "json"])
        if self.model:
            cmd.extend(["--model", self.model])
        if self.allowed_tools:
            for tool in self.allowed_tools:
                cmd.extend(["--allowedTools", tool])
        cmd.extend(["--max-turns", str(self.max_turns)])
        return cmd

    def run(self, prompt: str) -> str:
        """Run a prompt through claude -p and return the result text.

        Args:
            prompt: The prompt to send to claude.

        Returns:
            The result text from claude's response.

        Raises:
            RuntimeError: If the subprocess fails or times out.
        """
        cmd = self._build_command(prompt, streaming=False)
        logger.info("Running headless: %s", " ".join(cmd[:4]) + " ...")

        try:
            result = subprocess.run(
                cmd,
                capture_output=True,
                text=True,
                timeout=self.timeout,
            )
        except subprocess.TimeoutExpired:
            raise RuntimeError(
                f"HeadlessRunner timed out after {self.timeout}s"
            )
        except FileNotFoundError:
            raise RuntimeError(
                f"Claude binary not found: {self.claude_binary}. "
                "Ensure Claude CLI is installed and in PATH."
            )

        if result.returncode != 0:
            logger.error("HeadlessRunner stderr: %s", result.stderr)
            raise RuntimeError(
                f"HeadlessRunner exited with code {result.returncode}: {result.stderr[:500]}"
            )

        try:
            parsed = json.loads(result.stdout)
            return parsed.get("result", result.stdout)
        except json.JSONDecodeError:
            return result.stdout.strip()

    def run_streaming(self, prompt: str) -> list[dict[str, Any]]:
        """Run a prompt in streaming mode, collecting all JSON events.

        Args:
            prompt: The prompt to send to claude.

        Returns:
            List of parsed JSON event dicts from the stream.

        Raises:
            RuntimeError: If the subprocess fails or times out.
        """
        cmd = self._build_command(prompt, streaming=True)
        logger.info("Running headless (streaming): %s", " ".join(cmd[:4]) + " ...")

        try:
            proc = subprocess.Popen(
                cmd,
                stdout=subprocess.PIPE,
                stderr=subprocess.PIPE,
                text=True,
            )
        except FileNotFoundError:
            raise RuntimeError(
                f"Claude binary not found: {self.claude_binary}. "
                "Ensure Claude CLI is installed and in PATH."
            )

        events: list[dict[str, Any]] = []
        try:
            assert proc.stdout is not None
            for line in proc.stdout:
                line = line.strip()
                if not line:
                    continue
                try:
                    event = json.loads(line)
                    events.append(event)
                    logger.debug("Stream event: %s", event.get("type", "unknown"))
                except json.JSONDecodeError:
                    logger.warning("Non-JSON line in stream: %s", line[:200])
        finally:
            proc.wait(timeout=self.timeout)

        if proc.returncode != 0:
            stderr_output = proc.stderr.read() if proc.stderr else ""
            raise RuntimeError(
                f"HeadlessRunner (streaming) exited with code {proc.returncode}: "
                f"{stderr_output[:500]}"
            )

        return events

    def check_available(self) -> bool:
        """Check if the claude CLI is available.

        Returns:
            True if `claude --version` succeeds, False otherwise.
        """
        try:
            result = subprocess.run(
                [self.claude_binary, "--version"],
                capture_output=True,
                text=True,
                timeout=10,
            )
            return result.returncode == 0
        except (subprocess.TimeoutExpired, FileNotFoundError):
            return False
